From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Robert Beckett <bob.beckett@collabora.com>
Date: Tue, 6 Feb 2024 18:52:43 +0000
Subject: [PATCH] Close the FB if possible to keep the image on the screen

Add a new optional active retain handler for renderer plugins.
Drm requires active behaviour to mark output images to be
retained and so can not just rely on lack of cleanup to retain
the splash.

Signed-off-by: Robert Beckett <bob.beckett@collabora.com>
[seilent: rebased onto 24.004.60 with the Fedora f44 patches]
---
 src/libply-splash-core/ply-device-manager.c  | 17 ++++++++++
 src/libply-splash-core/ply-device-manager.h  |  1 +
 src/libply-splash-core/ply-renderer-plugin.h |  1 +
 src/libply-splash-core/ply-renderer.c        |  7 ++++
 src/libply-splash-core/ply-renderer.h        |  1 +
 src/main.c                                   | 11 ++++++-
 src/plugins/renderers/drm/plugin.c           | 34 ++++++++++++++++++++
 7 files changed, 71 insertions(+), 1 deletion(-)

diff --git a/src/libply-splash-core/ply-device-manager.c b/src/libply-splash-core/ply-device-manager.c
index 499d326..021debf 100644
--- a/src/libply-splash-core/ply-device-manager.c
+++ b/src/libply-splash-core/ply-device-manager.c
@@ -1343,6 +1343,23 @@ ply_device_manager_deactivate_renderers (ply_device_manager_t *manager)
         manager->renderers_activated = false;
 }
 
+static void
+retain_splash (char *device_path,
+               ply_renderer_t *renderer,
+               ply_device_manager_t *manager)
+{
+        ply_renderer_retain(renderer);
+}
+
+void
+ply_device_manager_retain_splash (ply_device_manager_t *manager)
+{
+        ply_hashtable_foreach (manager->renderers,
+                               (ply_hashtable_foreach_func_t *)
+                               retain_splash,
+                               manager);
+}
+
 void
 ply_device_manager_activate_keyboards (ply_device_manager_t *manager)
 {
diff --git a/src/libply-splash-core/ply-device-manager.h b/src/libply-splash-core/ply-device-manager.h
index fb5e1cf..7ad0ecb 100644
--- a/src/libply-splash-core/ply-device-manager.h
+++ b/src/libply-splash-core/ply-device-manager.h
@@ -76,6 +76,7 @@ void ply_device_manager_activate_keyboards (ply_device_manager_t *manager);
 void ply_device_manager_deactivate_keyboards (ply_device_manager_t *manager);
 void ply_device_manager_activate_renderers (ply_device_manager_t *manager);
 void ply_device_manager_deactivate_renderers (ply_device_manager_t *manager);
+void ply_device_manager_retain_splash (ply_device_manager_t *manager);
 ply_terminal_t *ply_device_manager_get_default_terminal (ply_device_manager_t *manager);
 
 #endif
diff --git a/src/libply-splash-core/ply-renderer-plugin.h b/src/libply-splash-core/ply-renderer-plugin.h
index 7d18f11..d7e8c75 100644
--- a/src/libply-splash-core/ply-renderer-plugin.h
+++ b/src/libply-splash-core/ply-renderer-plugin.h
@@ -84,6 +84,7 @@ typedef struct
                                  ply_input_device_t     *input_device);
         void (*remove_input_device)(ply_renderer_backend_t *backend,
                                     ply_input_device_t     *input_device);
+        void (*retain)(ply_renderer_backend_t *backend);
 } ply_renderer_plugin_interface_t;
 
 #endif /* PLY_RENDERER_PLUGIN_H */
diff --git a/src/libply-splash-core/ply-renderer.c b/src/libply-splash-core/ply-renderer.c
index 21df22f..21f6039 100644
--- a/src/libply-splash-core/ply-renderer.c
+++ b/src/libply-splash-core/ply-renderer.c
@@ -311,6 +311,13 @@ ply_renderer_close (ply_renderer_t *renderer)
         renderer->is_active = false;
 }
 
+void
+ply_renderer_retain (ply_renderer_t *renderer)
+{
+    if (renderer->plugin_interface->retain)
+        renderer->plugin_interface->retain(renderer->backend);
+}
+
 bool
 ply_renderer_handle_change_event (ply_renderer_t *renderer)
 {
diff --git a/src/libply-splash-core/ply-renderer.h b/src/libply-splash-core/ply-renderer.h
index 7c3406e..49e8e8b 100644
--- a/src/libply-splash-core/ply-renderer.h
+++ b/src/libply-splash-core/ply-renderer.h
@@ -59,6 +59,7 @@ void ply_renderer_free (ply_renderer_t *renderer);
 bool ply_renderer_open (ply_renderer_t *renderer,
                         bool            force);
 void ply_renderer_close (ply_renderer_t *renderer);
+void ply_renderer_retain (ply_renderer_t *renderer);
 /* Returns true when the heads have changed as a result of the change event */
 bool ply_renderer_handle_change_event (ply_renderer_t *renderer);
 void ply_renderer_activate (ply_renderer_t *renderer);
diff --git a/src/main.c b/src/main.c
index 4bb3c4b..ef28d46 100644
--- a/src/main.c
+++ b/src/main.c
@@ -1251,6 +1251,12 @@ hide_splash (state_t *state)
         }
 }
 
+static void
+retain_splash(state_t *state)
+{
+        ply_device_manager_retain_splash (state->device_manager);
+}
+
 static void
 dump_details_and_quit_splash (state_t *state)
 {
@@ -1342,7 +1348,10 @@ on_boot_splash_idle (state_t *state)
          * quit command, the quit command takes precedence.
          */
         if (state->quit_trigger != NULL) {
-                if (!state->should_retain_splash) {
+                if (state->should_retain_splash) {
+                        ply_trace ("retaining splash");
+                        retain_splash (state);
+                } else {
                         ply_trace ("hiding splash");
                         hide_splash (state);
                 }
diff --git a/src/plugins/renderers/drm/plugin.c b/src/plugins/renderers/drm/plugin.c
index f7766a5..bb94127 100644
--- a/src/plugins/renderers/drm/plugin.c
+++ b/src/plugins/renderers/drm/plugin.c
@@ -777,6 +777,24 @@ ply_renderer_head_unmap (ply_renderer_backend_t *backend,
         head->scan_out_buffer_id = 0;
 }
 
+static void
+ply_renderer_head_retain (ply_renderer_backend_t *backend,
+                          ply_renderer_head_t *head)
+{
+        int ret;
+
+        ply_renderer_buffer_t *buffer;
+
+        buffer = ply_hashtable_lookup (backend->output_buffers,
+                                       (void *) (uintptr_t) head->scan_out_buffer_id);
+        assert (buffer != NULL);
+
+        ply_trace("Closing FB %u (retaining)", buffer->id);
+        ret = drmModeCloseFB (backend->device_fd, buffer->id);
+        if (ret)
+                ply_trace("Could not close FB %u: %m. Attempting rm", buffer->id);
+}
+
 static void
 ply_renderer_head_remove (ply_renderer_backend_t *backend,
                           ply_renderer_head_t    *head)
@@ -1688,6 +1706,21 @@ unmap_from_device (ply_renderer_backend_t *backend)
         }
 }
 
+static void
+retain (ply_renderer_backend_t *backend)
+{
+        ply_renderer_head_t *head;
+        ply_list_node_t *node;
+
+        node = ply_list_get_first_node (backend->heads);
+        while (node != NULL) {
+                head = (ply_renderer_head_t *) ply_list_node_get_data (node);
+                ply_renderer_head_retain (backend, head);
+                node = ply_list_get_next_node (backend->heads, node);
+        }
+}
+
+
 static bool
 reset_scan_out_buffer_if_needed (ply_renderer_backend_t *backend,
                                  ply_renderer_head_t    *head)
@@ -2142,6 +2175,7 @@ ply_renderer_backend_get_interface (void)
                 .get_keymap                   = get_keymap,
                 .add_input_device             = add_input_device,
                 .remove_input_device          = remove_input_device,
+                .retain                       = retain,
         };
 
         return &plugin_interface;
-- 
2.55.0

